src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py

```
@@ -336,69 +334,71 @@ def _check_remote_source_code(source_location):
-                ".dockerignore: no rule for '%s'.", tarinfo.name)
-            return tarinfo
+    logger.debug("Starting to upload the archived source code from '%s'.", tar_file_path)
 
-        with tarfile.open(tar_file_path, "w:gz") as tar:
-            # NOTE: Need to set arcname to empty string;
-            # otherwise the child item name will have a prefix (eg, ../) which can block unpacking.
-            tar.add(source_location, arcname="", filter=_filter_file)
+    upload_url = None
+    error_message = "Could not get build source upload URL."
+    try:
+        source_upload_location = client.get_build_source_upload_url(resource_group_name, registry_name)
+        upload_url = source_upload_location.upload_url
+    except (AttributeError, CloudError) as e:
+        logger.debug("%s Exception: %s", error_message, e)
+        raise CLIError(error_message)
 
-        logger.debug(
-            "Starting to upload the archived source code from '%s'.", tar_file_path)
+    if not upload_url:
+        logger.debug("%s Empty build source upload URL.", error_message)
+        raise CLIError(error_message)
 
-        account_name, endpoint_suffix, container_name, blob_name, sas_token = _get_blob_info(
-            source_upload_location.upload_url)
+    account_name, endpoint_suffix, container_name, blob_name, sas_token = _get_blob_info(upload_url)
 
-        BlockBlobService(account_name=account_name, sas_token=sas_token,
+    try:
+        BlockBlobService(account_name=account_name,
+                         sas_token=sas_token,
                          endpoint_suffix=endpoint_suffix).create_blob_from_path(
-                             container_name=container_name, blob_name=blob_name, file_path=tar_file_path)
+                             container_name=container_name,
+                             blob_name=blob_name,
+                             file_path=tar_file_path)
 
         return source_upload_location.relative_path
     except Exception as err:
```

Still need more investigation what exception `BlockBlobService` might return to get rid of this bare-except ... /cc @ehotinger @northtyphoon

---

How do people deal with upgrading libraries in the CLI if they don't do bare exceptions for some cases? It's honestly crazy to me, what happens if some random library starts throwing `FooException` for a case like this? Whatcha gonna do... republish the CLI once a bunch of people have a bug?

---

https://github.com/Azure/azure-storage-python/blob/19d845d58551bd36a04c46628dc42089ee269362/azure-storage-blob/azure/storage/blob/blockblobservice.py

`ValueError` is one of them... but it requires munging through the `perform_request` code to get the final result

---

I might misunderstand. What is the bare-except? Are you referring `except Exception as err:` ? Is it anti-pattern?

---

The idea is that you want to catch the expected errors and handle them rather than catching everything because `except` or `except Exception` will hide the actual errors. Client side app doesn't have the luxury of various telemetry/logs to help diagnose so what might happen is an unexpected exception is thrown and we hide it while the command is not working.

@ehotinger Azure management plane has been pretty consistent in what exception they throw. In case of new exception types, chances are (1) if it inherits an existing exception, it's still fine, whoever catches the parent exception can still catch the child exception; or (2) if it's a brand new exception, it's just like a breaking change, the question then is, do you want to hide it when it happens or expose it, since it is not an expected error.

I do agree that data-plane SDKs are not as consistent as management SDKs so this is really a best effort practice.

---

Looking through this file more closely. It seems that the exception we are catching out of `blob_service` is `AzureHttpError` such as https://github.com/AzureCR/azure-cli/blob/dev/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py#L132 and https://github.com/AzureCR/azure-cli/blob/dev/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build.py#L148

---

@djyou just do whatever to appeal to the `azure-cli` PR gods I guess. It should be totally fine to `catch Exception`. Shouldn't be an anti-pattern; Python's own source code has plenty of it.